typing/typecore.ml

```diff
@@ -4032,15 +4035,30 @@ and type_expect_
               let error = Expr_not_a_record_type exp.exp_type in
               raise (Error (exp.exp_loc, env, error))
         in
+        (* In record update expressions [{r with ...}],
+           [r] may have an inline record type with a restricted scope.
+           We must propagate this restricted scope if we build a new
+           return type below. *)
+        let ty_record_scope = match opt_exp with
+          | None -> Ident.lowest_scope
+          | Some exp -> get_scope exp.exp_type
+        in
         match expected_opath, opt_exp_opath with
-        | None, None -> newvar (), None
+        | None, None ->
+            let ty = newvar () in
+            update_scope ty_record_scope ty;
+            ty, None
         | Some _, None -> ty_expected, expected_opath
         | Some(_, _, true), Some _ -> ty_expected, expected_opath
         | (None | Some (_, _, false)), Some (_, p', _) ->
             let decl = Env.find_type p' env in
             let ty =
               with_local_level_generalize_structure
-                (fun () -> newconstr p' (instance_list decl.type_params))
+                (fun () ->
+                   let ty = newconstr p' (instance_list decl.type_params) in
+                   update_scope ty_record_scope ty ;
```

Comment:
Same thing here.
Also, it looks like this actually allows:
```ocaml
fun (A r) -> let r' = {r with x = ..} in A r
```
But then what about
```ocaml
fun (A r) -> (fun r' -> A r') {r with x = ..}
```